syntax = "proto3";

package muicore;

service ManticoreUI {
    rpc StartNative(NativeArguments) returns (ManticoreInstance) {}
    rpc StartEVM(EVMArguments) returns (ManticoreInstance) {}
    rpc Terminate(ManticoreInstance) returns (TerminateResponse) {}
    rpc GetStateList(ManticoreInstance) returns (MUIStateList) {}
    rpc GetMessageList(ManticoreInstance) returns (MUIMessageList) {}
    rpc CheckManticoreRunning(ManticoreInstance) returns (ManticoreRunningStatus) {}
    rpc TargetAddressNative(AddressRequest) returns (TargetResponse) {}
    rpc StopServer(StopServerRequest) returns (StopServerResponse) {}
}

// LogMessage and StateList message types have "MUI" in their names to distinguish them from those in mserialize

message MUILogMessage{ 
    string content = 1;
}

message MUIMessageList{
    repeated MUILogMessage messages = 2; 
}

message MUIState {
    int32 state_id = 3;
    uint64 pc = 10;
}

message MUIStateList{

    // state categories in MUI - based on manticore enums StateStatus and StateList
    repeated MUIState active_states = 4;
    repeated MUIState waiting_states = 5;
    repeated MUIState forked_states = 6;
    repeated MUIState errored_states = 7;
    repeated MUIState complete_states = 8;
}

message ManticoreInstance {
    string uuid = 9;
}

message TerminateResponse {}

message NativeArguments {
    string program_path = 11;
    repeated string binary_args = 16;    
    repeated string envp = 17;
    repeated string symbolic_files = 18;
    string concrete_start = 19;
    string stdin_size = 20;
    string additional_mcore_args = 21;
}

message EVMArguments {
    string contract_path = 12;
    string contract_name = 13;
    string solc_bin = 14;
    string tx_limit = 22;
    string tx_account = 23;
    repeated string detectors_to_exclude = 24;
    string additional_flags = 25;
}

message AddressRequest {

    enum TargetType {
        FIND = 0;
        AVOID = 1;
        CLEAR = 2;
    }

    uint64 address = 26;
    TargetType type = 27;
}

message TargetResponse {}

message ManticoreRunningStatus {
    bool is_running = 15;
}

message StopServerRequest {}

message StopServerResponse {}